test/unit/master/test_cluster_master.py

```diff
@@ -213,6 +213,38 @@ def test_updating_slave_to_nonexistent_state_should_raise_bad_request_error(self
         with self.assertRaises(BadRequestError):
             master.handle_slave_state_update(slave, 'NONEXISTENT_STATE')
 
+    def test_async_receive_heartbeat_from_slave_calls_set_heartbeat_for_the_slave(self):
+        master = ClusterMaster()
+        slave_url = "url"
+        slave_id = master.connect_slave(slave_url, 1)
+        slave = Mock()
+        master.get_slave = MagicMock(return_value=slave)
+        master._async_receive_heartbeat_from_slave(slave_id['slave_id'])
+        master.get_slave.assert_called_with(slave_id['slave_id'])
+        self.assertEqual(slave.set_heartbeat.call_count, 1, 'incoming heartbeat sets timestamp for correct slave')
+
+    @genty_dataset (
+        slave_unresponsive=(True,False,),
+        slave_dead=(False,True,),
+        slave_responsive=(True,True,),
+    )
+    def test_heartbeat_disconnects_unresponsive_slave(self, slave_alive, slave_responsive):
+        master = ClusterMaster()
+        slave_url = "url"
+        slave = master.connect_slave(slave_url, 1)
+        slave = master.get_slave(int(slave['slave_id']))
+
+        slave.is_alive = MagicMock(return_value=slave_alive)
```

Comment:
Same comment about clobbering a method from a real object.

Instead of clobbering `slave.is_alive`, use patch to patch out the `Slave()` constructor, that way you have a reference to a mock slave object, and set the attributes on that. 

Instead of clobbering master._disconnect_slave and asserting that that's called, patch the Slave constructor and, assert `slave.mark_dead()` is called